### PR TITLE
Enrich status with HTML Microformats

### PR DESCRIPTION
**Pull request by xtuc**

### CODE REVIEW

**Review by dario-piotrowicz — APPROVED** [review]
very nice! 😃 

(nitpick: I haven't tried but I get the feeling that a `enrichStatus` could be implemented more cleaning using regex substitutions)

File: backend/src/mastodon/microformats.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		if (char === '@') {
+			state = 'mention'
+		}
+
+		if (
+			char === 'h' &&
+			status[i + 1] === 't' &&
+			status[i + 2] === 't' &&
+			status[i + 3] === 'p' &&
+			status[i + 4] === ':'
```
[the remark this answers is not in the document]
Comment (xtuc) [reply backend/src/mastodon/microformats.ts:29]: Oh, nice. Thanks!

File: backend/src/mastodon/microformats.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		) {
+			state = 'link'
+		}
+		if (
```
[the remark this answers is not in the document]
Comment (xtuc) [reply backend/src/mastodon/microformats.ts:33]: It's easier to run when there's line in between the conditions IMO

File: backend/src/mastodon/microformats.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+function tag(name: string, content: string, attrs: object = {}): string {
+	let htmlAttrs = ''
+	for (const [key, value] of Object.entries(attrs)) {
+		htmlAttrs += ` ${key}="${value}"`
+	}
+
+	return `<${name}${htmlAttrs}>${content}</${name}>`
+}
+
+/// Transform a text status into a HTML status; enriching it with links / mentions.
```
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply backend/src/mastodon/microformats.ts:12]: I just personally think it is nicer and looks more polished, but yeah it's not like super important here 🙂

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply backend/src/mastodon/microformats.ts:12]: Interestingly triple slashes in TS can have special meaning, sometimes... https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html

But the main benefit of this being a JSDOC style comment (`/** ... */`) is that it then appears in the intellisense.

File: backend/test/mastodon.spec.ts [hunk 4]
```diff
@@ -281,4 +282,21 @@ describe('Mastodon APIs', () => {
...
+
+	describe('Microformats', () => {
+		test('convert mentions to HTML', () => {
+			const status = 'hey @test@example.com hi'
+
+			assert.equal(
+				enrichStatus(status),
+				'<p>hey <span class="h-card"><a href="https://example.com/@test" class="u-url mention">@<span>test</span></a></span> hi</p>'
+			)
+		})
```
Comment (petebacondarwin) [inline backend/test/mastodon.spec.ts:294]: Is it worth testing false positives? E.g. if something had already converted a mention to an anchor elsewhere? Do we handle that or just double enrich it?

File: backend/src/mastodon/microformats.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (xtuc) [reply backend/src/mastodon/microformats.ts:12]: Oops, it should just be double slash. Tripple slashes is Rust muscle memory.

File: backend/test/mastodon.spec.ts (hunk 4 above)
Comment (xtuc) [reply backend/test/mastodon.spec.ts:294]: Links aren't safe from double enriching. We could just check for the first <p> tag to see if it's already HTML or not.

**Pull request merged**